src/coprocessor/dag/rpn_expr/types.rs

```diff
@@ -0,0 +1,556 @@
+            RpnStackNode::Vector { .. } => panic!(),
+        }
+    }
+
+    #[inline]
+    pub fn vector_value(&self) -> VectorValueRef {
+        match self {
+            RpnStackNode::Scalar { .. } => panic!(),
+            RpnStackNode::Vector { ref value, .. } => value.borrow(),
+        }
+    }
+
+    #[inline]
+    pub fn is_scalar(&self) -> bool {
+        match self {
+            RpnStackNode::Scalar { .. } => true,
+            _ => false,
+        }
+    }
+
+    #[inline]
+    pub fn is_vector(&self) -> bool {
+        match self {
+            RpnStackNode::Vector { .. } => true,
+            _ => false,
+        }
+    }
+}
+
+/// Type for each node in the RPN expression list.
+#[derive(Debug)]
+pub enum RpnExpressionNode {
+    /// Represents a function.
+    Fn {
+        func: RpnFunction,
+        field_type: FieldType,
+    },
+
+    /// Represents a scalar constant value.
+    Constant {
+        value: ScalarValue,
+        field_type: FieldType,
+    },
+
+    /// Represents a reference to a table column.
+    TableColumnRef {
+        offset: usize,
+
+        // Although we can know `ColumnInfo` according to `offset` and columns info in scan
+        // executors, its type is `ColumnInfo` instead of `FieldType`..
+        field_type: FieldType,
+    },
+}
+
+impl RpnExpressionNode {
+    #[inline]
+    pub fn field_type(&self) -> &FieldType {
+        match self {
+            RpnExpressionNode::Fn { ref field_type, .. } => field_type,
+            RpnExpressionNode::Constant { ref field_type, .. } => field_type,
+            RpnExpressionNode::TableColumnRef { ref field_type, .. } => field_type,
+        }
+    }
+
+    #[inline]
+    pub fn fn_func(&self) -> Option<RpnFunction> {
+        match self {
+            RpnExpressionNode::Fn { ref func, .. } => Some(*func),
+            _ => None,
+        }
+    }
+
+    #[inline]
+    pub fn constant_value(&self) -> Option<&ScalarValue> {
+        match self {
+            RpnExpressionNode::Constant { ref value, .. } => Some(value),
+            _ => None,
+        }
+    }
+
+    #[inline]
+    pub fn table_column_ref_offset(&self) -> Option<usize> {
+        match self {
+            RpnExpressionNode::TableColumnRef { ref offset, .. } => Some(*offset),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct RpnExpressionNodeVec(Vec<RpnExpressionNode>);
+
+impl ::std::ops::Deref for RpnExpressionNodeVec {
+    type Target = Vec<RpnExpressionNode>;
+
+    fn deref(&self) -> &Vec<RpnExpressionNode> {
+        &self.0
+    }
+}
+
+impl ::std::ops::DerefMut for RpnExpressionNodeVec {
+    fn deref_mut(&mut self) -> &mut Vec<RpnExpressionNode> {
+        &mut self.0
+    }
+}
+
+impl RpnExpressionNodeVec {
+    pub fn build_from_def(def: Expr, time_zone: Tz) -> Self {
+        let mut expr_nodes = Vec::new();
+        Self::append_rpn_nodes_recursively(def, time_zone, &mut expr_nodes);
+        expr_nodes.reverse();
```

Comment:
Normally RPN expressions evaluate from left to right in sequence. If we don't reverse it, we have to evaluate it from right to left, which seems to be more confusing.

I didn't add comment for the reverse part because explaining why there is a reverse is too complicated. Maybe we can leave a TODO here to improve the comment in future?